pkg/v1/remote/write_test.go

```diff
@@ -606,3 +608,139 @@ func TestWriteWithErrors(t *testing.T) {
+	if err != nil {
+		t.Fatalf("name.NewTag() = %v", err)
+	}
+
+	anotherRepo2, err := name.NewTag(fmt.Sprintf("%s/%s:latest", u.Host, "sample/another_repo2"), name.WeakValidation)
+	if err != nil {
+		t.Fatalf("name.NewTag() = %v", err)
+	}
+
+	img := setupImage(t)
+	layers, err := img.Layers()
+	if err != nil {
+		t.Fatalf("img.Layers() = %v", err)
+	}
+	dummyLayer := layers[0]
+
+	testCases := []struct {
+		name      string
+		reference name.Reference
+		layers    []v1.Layer
+		expected  []string
+	}{
+		{
+			name:      "empty layers",
+			reference: referenceToUpload,
+			layers:    []v1.Layer{},
+			expected: []string{
+				referenceToUpload.Scope(transport.PushScope),
+			},
+		},
+		{
+			name:      "mountable layers with single reference with no duplicate",
+			reference: referenceToUpload,
+			layers: []v1.Layer{
+				&MountableLayer{
+					Layer:     dummyLayer,
+					Reference: anotherRepo1,
+				},
+			},
+			expected: []string{
+				referenceToUpload.Scope(transport.PushScope),
+				anotherRepo1.Scope(transport.PullScope),
+			},
+		},
+		{
+			name:      "mountable layers with single reference with duplicate",
+			reference: referenceToUpload,
+			layers: []v1.Layer{
+				&MountableLayer{
+					Layer:     dummyLayer,
+					Reference: anotherRepo1,
+				},
+				&MountableLayer{
+					Layer:     dummyLayer,
+					Reference: anotherRepo1,
+				},
+			},
+			expected: []string{
+				referenceToUpload.Scope(transport.PushScope),
+				anotherRepo1.Scope(transport.PullScope),
+			},
+		},
+		{
+			name:      "mountable layers with multiple references with no-duplicates",
+			reference: referenceToUpload,
+			layers: []v1.Layer{
+				&MountableLayer{
+					Layer:     dummyLayer,
+					Reference: anotherRepo1,
+				},
+				&MountableLayer{
+					Layer:     dummyLayer,
+					Reference: anotherRepo2,
+				},
+			},
+			expected: []string{
+				referenceToUpload.Scope(transport.PushScope),
+				anotherRepo1.Scope(transport.PullScope),
+				anotherRepo2.Scope(transport.PullScope),
+			},
+		},
+		{
+			name:      "mountable layers with multiple references with duplicates",
+			reference: referenceToUpload,
+			layers: []v1.Layer{
+				&MountableLayer{
+					Layer:     dummyLayer,
+					Reference: anotherRepo1,
+				},
+				&MountableLayer{
+					Layer:     dummyLayer,
+					Reference: anotherRepo2,
+				},
+				&MountableLayer{
+					Layer:     dummyLayer,
+					Reference: anotherRepo1,
+				},
+				&MountableLayer{
+					Layer:     dummyLayer,
+					Reference: anotherRepo2,
+				},
+			},
+			expected: []string{
+				referenceToUpload.Scope(transport.PushScope),
+				anotherRepo1.Scope(transport.PullScope),
+				anotherRepo2.Scope(transport.PullScope),
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		actual := scopesForUploadingImage(testCase.reference, testCase.layers)
+		if !reflect.DeepEqual(sort.StringsAreSorted(actual), sort.StringsAreSorted(testCase.expected)) {
```

Comment:
I think you want `sort.Strings` instead of `sort.StringsAreSorted`, but I'd probably use the [`cmpopts.SortSlices`](https://godoc.org/github.com/google/go-cmp/cmp/cmpopts#SortSlices) instead.

Per my other comment, I think you want to assert that the first element is what you expect (the push scope)... something like this:
```go
if want, got := testCase.expected[0], actual[0]; want != got {
  t.Errorf("Wrong first scope; want %v, got %v", want, got)
}
if diff := cmp.Diff(testCase.expected[1:], actual[1:], cmpopts.SortSlices()); diff != "" {
  t.Errorf("Wrong scopes (-want +got) = %v", diff)
}
```

I haven't tested that that even compiles, but it should be approximately right :P